File: t/VB/OcrMultiEngineDemo/ViewerControl/ZoneAnnotationObjectRenderer.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Runtime.Serialization
Imports System.Drawing
Imports System.Drawing.Drawing2D

Imports Leadtools
Imports Leadtools.Ocr
Imports Leadtools.Annotations.Engine
Imports Leadtools.Annotations.Designers
Imports Leadtools.Annotations.Rendering
Imports Leadtools.Drawing

Public Class ZoneAnnotationObjectRenderer : Inherits AnnRectangleObjectRenderer
   Private _ocrPage As IOcrPage
   Private _cellPen As AnnStroke

   Public Sub New()
      MyBase.New()
      _ocrPage = Nothing
   End Sub

   Public Property OcrPage() As IOcrPage
      Get
         Return _ocrPage
      End Get
      Set(value As IOcrPage)
         _ocrPage = value
      End Set
   End Property

   Public Property CellPen() As AnnStroke
      Get
         Return _cellPen
      End Get
      Set(ByVal value As AnnStroke)
         _cellPen = value
      End Set
   End Property

   Public Overrides Sub Render(mapper As AnnContainerMapper, annObject As AnnObject)
      Dim zoneObject As ZoneAnnotationObject = TryCast(annObject, ZoneAnnotationObject)
      If Not IsNothing(_ocrPage) AndAlso zoneObject.ZoneIndex >= 0 AndAlso zoneObject.ZoneIndex < _ocrPage.Zones.Count Then
         Dim engine As AnnWinFormsRenderingEngine = CType(RenderingEngine, AnnWinFormsRenderingEngine)
         If Not engine Is Nothing AndAlso Not engine.Context Is Nothing AndAlso Not zoneObject Is Nothing AndAlso Not zoneObject.Cells Is Nothing Then

            Dim graphics As Graphics = engine.Context
            Dim zone As OcrZone = _ocrPage.Zones(zoneObject.ZoneIndex)
            Dim cells As OcrZoneCell() = Nothing
            cells = _ocrPage.Zones.GetZoneCells(zone)
            If Not _ocrPage.TableZoneManager Is Nothing And zone.ZoneType = OcrZoneType.Table AndAlso Not cells Is Nothing AndAlso cells.Length > 0 AndAlso Not CellPen Is Nothing Then
               Dim gState As GraphicsState = graphics.Save()
               If Not gState Is Nothing Then
                  For Each cell As OcrZoneCell In zoneObject.Cells
                     Dim rc As LeadRectD = New LeadRectD(cell.Bounds.X, cell.Bounds.Y, cell.Bounds.Width, cell.Bounds.Height)
                     rc = mapper.RectFromContainerCoordinates(rc, AnnFixedStateOperations.None)
                     If (Not rc.IsEmpty) Then
                        ' Draw cells borders as lines in order not to draw the borders with 0 width
                        DrawLine(graphics, OcrCellBorder.Left, cell.LeftBorderStyle, cell.LeftBorderWidth, Convert.ToSingle(rc.Left), Convert.ToSingle(rc.Top), Convert.ToSingle(rc.Left), Convert.ToSingle(rc.Bottom))
                        DrawLine(graphics, OcrCellBorder.Top, cell.TopBorderStyle, cell.TopBorderWidth, Convert.ToSingle(rc.Left), Convert.ToSingle(rc.Top), Convert.ToSingle(rc.Right), Convert.ToSingle(rc.Top))
                        DrawLine(graphics, OcrCellBorder.Right, cell.RightBorderStyle, cell.RightBorderWidth, Convert.ToSingle(rc.Right), Convert.ToSingle(rc.Top), Convert.ToSingle(rc.Right), Convert.ToSingle(rc.Bottom))
                        DrawLine(graphics, OcrCellBorder.Bottom, cell.BottomBorderStyle, cell.BottomBorderWidth, Convert.ToSingle(rc.Left), Convert.ToSingle(rc.Bottom), Convert.ToSingle(rc.Right), Convert.ToSingle(rc.Bottom))
                     End If
                  Next cell
                  graphics.Restore(gState)
               End If
            End If
         End If
      End If
      MyBase.Render(mapper, annObject)
   End Sub

   Private Sub DrawLine(ByVal g As Graphics, ByVal border As OcrCellBorder, ByVal lineStyle As OcrCellBorderLineStyle, ByVal penWidth As Integer, ByVal x1 As Single, ByVal y1 As Single, ByVal x2 As Single, ByVal y2 As Single)
      If Not lineStyle = OcrCellBorderLineStyle.None AndAlso penWidth > 0 Then
         Using pen As Pen = New Pen(Color.FromName((CType(CellPen.Stroke, AnnSolidColorBrush)).Color))
            pen.Width = penWidth
            pen.EndCap = LineCap.Square
            pen.StartCap = LineCap.Square

            Select Case lineStyle
               Case OcrCellBorderLineStyle.Double
                  pen.DashStyle = DashStyle.Solid
                  Select Case border
                     Case OcrCellBorder.Left, OcrCellBorder.Right
                        g.DrawLine(pen, x1 - pen.Width, y1, x2 - pen.Width, y2)
                        g.DrawLine(pen, x1 + pen.Width, y1, x2 + pen.Width, y2)

                     Case OcrCellBorder.Top, OcrCellBorder.Bottom
                        g.DrawLine(pen, x1, y1 - pen.Width, x2, y2 - pen.Width)
                        g.DrawLine(pen, x1, y1 + pen.Width, x2, y2 + pen.Width)
                  End Select

               Case OcrCellBorderLineStyle.Dashed
                  pen.DashStyle = DashStyle.Dash

               Case OcrCellBorderLineStyle.Dotted
                  pen.DashStyle = DashStyle.Dot

               Case Else
                  pen.DashStyle = DashStyle.Solid
            End Select

            If Not lineStyle = OcrCellBorderLineStyle.Double Then
               g.DrawLine(pen, x1, y1, x2, y2)
            End If
         End Using
      Else
         ' Draw invisible cell border
         Using pen As Pen = New Pen(Color.FromName((CType(CellPen.Stroke, AnnSolidColorBrush)).Color))
            pen.Width = 1
            pen.DashStyle = DashStyle.Dot
            pen.Color = Color.FromArgb(16, Color.FromName((CType(CellPen.Stroke, AnnSolidColorBrush)).Color))
            g.DrawLine(pen, x1, y1, x2, y2)
         End Using
      End If
   End Sub
End Class

File: t/CS/OcrMultiEngineDemo/ViewerControl/ZoneAnnotationObjectRenderer.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.Serialization;
using System.Drawing;
using System.Drawing.Drawing2D;

using Leadtools;
using Leadtools.Ocr;
using Leadtools.Drawing;
using Leadtools.Annotations.Engine;
using Leadtools.Annotations.Designers;
using Leadtools.Annotations.Rendering;

namespace OcrMultiEngineDemo
{
   public class ZoneAnnotationObjectRenderer : AnnRectangleObjectRenderer
   {
      private IOcrPage _ocrPage;
      private AnnStroke _cellPen;

      public ZoneAnnotationObjectRenderer() :
         base()
      {
         _ocrPage = null;
      }


      public IOcrPage OcrPage
      {
         get { return _ocrPage; }
         set { _ocrPage = value; }
      }


      public AnnStroke CellPen
      {
         get { return _cellPen; }
         set { _cellPen = value; }
      }

      public override void Render(AnnContainerMapper mapper, AnnObject annObject)
      {
         ZoneAnnotationObject zoneObject = annObject as ZoneAnnotationObject;
         if (_ocrPage != null && zoneObject.ZoneIndex >= 0 && zoneObject.ZoneIndex < _ocrPage.Zones.Count)
         {
            AnnWinFormsRenderingEngine engine = RenderingEngine as AnnWinFormsRenderingEngine;
            
            if (engine != null && engine.Context != null && zoneObject != null && zoneObject.Cells != null)
            {
               Graphics graphics = engine.Context;
               OcrZone zone = _ocrPage.Zones[zoneObject.ZoneIndex];
               OcrZoneCell[] cells = null;
               cells = _ocrPage.Zones.GetZoneCells(zone);

               if (_ocrPage.TableZoneManager != null && zone.ZoneType == OcrZoneType.Table && cells != null && cells.Length > 0 && CellPen != null)
               {
                  GraphicsState gState = graphics.Save();
                  if (gState != null)
                  {
                     foreach (OcrZoneCell cell in zoneObject.Cells)
                     {
                        LeadRectD rc = new LeadRectD(cell.Bounds.X, cell.Bounds.Y, cell.Bounds.Width, cell.Bounds.Height);
                        rc = mapper.RectFromContainerCoordinates(rc, AnnFixedStateOperations.None);

                        if (!rc.IsEmpty)
                        {
                           // Draw cells borders as lines in order not to draw the borders with 0 width
                           DrawLine(graphics, OcrCellBorder.Left, cell.LeftBorderStyle, cell.LeftBorderWidth, (float)rc.Left, (float)rc.Top, (float)rc.Left, (float)rc.Bottom);
                           DrawLine(graphics, OcrCellBorder.Top, cell.TopBorderStyle, cell.TopBorderWidth, (float)rc.Left, (float)rc.Top, (float)rc.Right, (float)rc.Top);
                           DrawLine(graphics, OcrCellBorder.Right, cell.RightBorderStyle, cell.RightBorderWidth, (float)rc.Right, (float)rc.Top, (float)rc.Right, (float)rc.Bottom);
                           DrawLine(graphics, OcrCellBorder.Bottom, cell.BottomBorderStyle, cell.BottomBorderWidth, (float)rc.Left, (float)rc.Bottom, (float)rc.Right, (float)rc.Bottom);
                        }
                     }

                     graphics.Restore(gState);
                  }
               }
            }
         }

         base.Render(mapper, annObject);
      }

      void DrawLine(Graphics g, OcrCellBorder border, OcrCellBorderLineStyle lineStyle, int penWidth, float x1, float y1, float x2, float y2)
      {
         if (lineStyle != OcrCellBorderLineStyle.None && penWidth > 0)
         {
            using (Pen pen = new Pen(Color.FromName((CellPen.Stroke as AnnSolidColorBrush).Color)))
            {
               pen.Width = penWidth;
               pen.EndCap = LineCap.Square;
               pen.StartCap = LineCap.Square;

               switch (lineStyle)
               {
                  case OcrCellBorderLineStyle.Double:
                     pen.DashStyle = DashStyle.Solid;
                     switch (border)
                     {
                        case OcrCellBorder.Left:
                        case OcrCellBorder.Right:
                           g.DrawLine(pen, x1 - pen.Width, y1, x2 - pen.Width, y2);
                           g.DrawLine(pen, x1 + pen.Width, y1, x2 + pen.Width, y2);
                           break;

                        case OcrCellBorder.Top:
                        case OcrCellBorder.Bottom:
                           g.DrawLine(pen, x1, y1 - pen.Width, x2, y2 - pen.Width);
                           g.DrawLine(pen, x1, y1 + pen.Width, x2, y2 + pen.Width);
                           break;
                     }
                     break;

                  case OcrCellBorderLineStyle.Dashed:
                     pen.DashStyle = DashStyle.Dash;
                     break;

                  case OcrCellBorderLineStyle.Dotted:
                     pen.DashStyle = DashStyle.Dot;
                     break;

                  case OcrCellBorderLineStyle.Solid:
                  default:
                     pen.DashStyle = DashStyle.Solid;
                     break;
               }

               if (lineStyle != OcrCellBorderLineStyle.Double)
                  g.DrawLine(pen, x1, y1, x2, y2);
            }
         }
         else
         {
            // Draw invisible cell border
            using (Pen pen = new Pen(Color.FromName((CellPen.Stroke as AnnSolidColorBrush).Color)))
            {
               pen.Width = 1;
               pen.DashStyle = DashStyle.Dot;
               pen.Color = Color.FromArgb(16, Color.FromName((CellPen.Stroke as AnnSolidColorBrush).Color));
               g.DrawLine(pen, x1, y1, x2, y2);
            }
         }
      }

   }
}
